src/sql/query.rs

```diff
@@ -186,8 +186,14 @@ pub fn eval_as_of<'a>(scx: &'a StatementContext, expr: Expr) -> Result<Timestamp
         }
         ScalarType::Int32 => evaled.unwrap_int32().try_into()?,
         ScalarType::Int64 => evaled.unwrap_int64().try_into()?,
-        ScalarType::TimestampTz => evaled.unwrap_timestamptz().timestamp().try_into()?,
-        ScalarType::Timestamp => evaled.unwrap_timestamp().timestamp().try_into()?,
+        ScalarType::TimestampTz => {
+            let ts = evaled.unwrap_timestamptz();
+            (ts.timestamp() * 1000 + i64::from(ts.timestamp_subsec_millis())).try_into()?
```

Comment:
maybe consider wrapping this in a `milliseconds` function that you can call after `unwrap_timestamp*`?

Reply:
I thought about that, but I'm not sure I like it better, since there would be two almost identical versions for timestamp, which has a `NaiveDateTime`, and timestamptz, which has a `DateTime<Utc>` (unless there's some way to unify those?) and at that point I figured I'd rather just keep them inlined as close to each other as possible. I'm open to being convinced otherwise though.

Reply:
so the trait TimestampLike kind of unifies the two and I think you can make a new function for milliseconds there that has a default implementation that relies on `timestamp` and `timestamp_subsec_micros` which are both trait functions that are already implemented for the two types

Reply:
agree that everything is kind of ugly and hacky around these two types. I'm not convinced that we should have made them be two types in the first place

Reply:
nikhils suggestion works too -- both types define a `timestamp_millis` upstream (confusing _not_ a part of the TimeLike trait)

Reply:
Right, it's part of the `chrono::DateTime` trait which is their equivalent to our `Timelike` trait. Possibly we should call ours `MzDateTime` to make the correspondence clearer.

Reply:
wait sorry just to check my understanding `chrono::DateTime` is a struct right? `chrono::DateLike` and `chrono::TimeLike` are traits and then our trait is `TimestampLike`?

Reply:
Changed to use `timestamp_millis`!